### PR TITLE
[custom channels]: extract some independent commits from mega staging branch

### PR DESCRIPTION
**Pull request by guggero**
As requested in https://github.com/lightningnetwork/lnd/pull/8960, I'm pulling out some more or less independent commits out from the mega-PR.

### CODE REVIEW

File: cmd/commands/commands_test.go [hunk 1]
```diff
@@ -120,3 +120,57 @@ func TestParseTimeLockDelta(t *testing.T) {
 		}
 	}
 }
+
+// TestReplaceCustomData tests that hex encoded custom data can be formatted as
+// JSON in the console output.
+func TestReplaceCustomData(t *testing.T) {
```
Comment (yyforyongyu) [inline cmd/commands/commands_test.go:126]: nit: add a new test `TestReplaceCustomDataErr` to check the error case?

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thanks a lot for the PR! Indeed very easy to review, just a few comments and nits.

File: lnwire/custom_records.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+// which must be greater than or equal to MinCustomRecordsTlvType.
+type CustomRecords map[uint64][]byte
+
+// NewCustomRecords creates a new CustomRecords instance from a
+// tlv.TypeMap.
+func NewCustomRecords(tlvMap tlv.TypeMap) (CustomRecords, error) {
+	// Make comparisons in unit tests easy by returning nil if the map is
+	// empty.
+	if len(tlvMap) == 0 {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/custom_records.go:27]: This makes a lot of the code where we actually use this much nicer. So we don't need to check for `len()` every time we create the custom records from the remaining parsed TLV fields.
It also simplifies some of the unit tests, in that encoding/decoding of an empty custom record produces the same result, which is `nil` (instead of being an empty map after decoding).

How strongly do you feel about this? Changing it will impact many commits in the mega-PR...

File: lnwire/custom_records.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		return nil, fmt.Errorf("error decoding HTLC record: %w", err)
+	}
+
+	return NewCustomRecords(typeMap)
+}
+
+// Validate checks that all custom records are in the custom type range.
+func (c CustomRecords) Validate() error {
+	if c == nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/custom_records.go:63]: See comment above.

File: funding/manager.go [hunk 4]
```diff
@@ -798,9 +798,13 @@ func (f *Manager) rebroadcastFundingTx(c *channeldb.OpenChannel) {
 	}
 }
 
+// PendingChanID is a type that represents a pending channel ID. This might be
+// selected by the caller, but if not, will be automatically selected.
+type PendingChanID = [32]byte
```
[the remark this answers is not in the document]
Comment (guggero) [reply funding/manager.go:806]: Hmm, I'm not sure I understand what you mean? This is a type definition, not a variable (so not a zero pending channel ID but a type shortcut).

File: cmd/commands/commands.go [hunk 5]
```diff
@@ -41,8 +42,46 @@ const (
...
+
+	customDataPattern = regexp.MustCompile(
+		`"custom_channel_data":\s*"([0-9a-z]+)"`,
+	)
+)
+
+// replaceCustomData replaces the custom channel data hex string with the
+// decoded custom channel data in the JSON response.
+func replaceCustomData(jsonBytes []byte) ([]byte, error) {
+	if customDataPattern.Match(jsonBytes) {
```
[the remark this answers is not in the document]
Comment (guggero) [reply cmd/commands/commands.go:58]: Good idea, done.

File: lnwire/custom_records.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	customRecords := make(CustomRecords, len(tlvMap))
+	for k, v := range tlvMap {
+		customRecords[uint64(k)] = v
+	}
+
+	// Validate the custom records.
+	err := customRecords.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("custom records from tlv map "+
+			"validation error: %v", err)
```
[the remark this answers is not in the document]
Comment (guggero) [reply lnwire/custom_records.go:39]: Nice catch, fixed. Not sure if we've turned on that linter in `lnd`?

[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lnwire/custom_records.go:39]: yeah i've seen this linter failure before...hmm

File: lnwire/custom_records.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lnwire/custom_records.go:27]: just somewhat, but non-blocking, def don't wanna bring large rebase conflicts and I can always create a PR if I reeeeally feel that strong

File: cmd/commands/commands_test.go (hunk 1 above)
Comment (guggero) [reply cmd/commands/commands_test.go:126]: Good idea! Added some negative test cases instead.

File: funding/manager.go (hunk 4 above)
Comment (ProofOfKeags) [inline funding/manager.go:806]: Why not just use ChanID? afaict they are representationally equivalent and because you went for a type alias (as opposed to a wrapper type) this doesn't create any additional domain type safety.

File: cmd/commands/main.go [hunk 7]
```diff
@@ -0,0 +1,601 @@
+// Copyright (c) 2013-2017 The btcsuite developers
+// Copyright (c) 2015-2016 The Decred developers
+// Copyright (C) 2015-2022 The Lightning Network Developers
+
+package commands
```
Comment (ProofOfKeags) [inline cmd/commands/main.go:5]: Do you want to do a git mv on this file as opposed to the current delete + create?

File: input/script_utils.go [hunk 8]
```diff
@@ -2296,9 +2311,6 @@ func TaprootCommitSpendSuccess(signer Signer, signDesc *SignDescriptor,
 	witnessStack[0] = maybeAppendSighash(sweepSig, signDesc.HashType)
 	witnessStack[1] = signDesc.WitnessScript
 	witnessStack[2] = ctrlBlockBytes
-	if err != nil {
-		return nil, err
-	}
```
Comment (ProofOfKeags) [inline input/script_utils.go:2301]: lol

File: cmd/commands/arg_parse.go [hunk 9]
```diff
@@ -42,7 +42,7 @@ func parseTime(s string, base time.Time) (uint64, error) {
 
 var lightningPrefix = "lightning:"
 
-// stripPrefix removes accidentally copied 'lightning:' prefix.
-func stripPrefix(s string) string {
+// StripPrefix removes accidentally copied 'lightning:' prefix.
+func StripPrefix(s string) string {
```
Comment (ProofOfKeags) [inline cmd/commands/arg_parse.go:46]: Seems odd to export this utility function.

**Review by ProofOfKeags — APPROVED** [review]

File: funding/manager.go (hunk 4 above)
Comment (Roasbeef) [reply funding/manager.go:806]: Alias was useful in `tapd`, so we can reference this value. `ChanID` is derived from a `wire.OutPoint`, while this is just a random value.

**Review by yyforyongyu — APPROVED** [review]
LGTM - the send coverage CI seems broken for this build for unknown reason, the rest are flakes.

File: cmd/commands/main.go (hunk 7 above)
Comment (guggero) [reply cmd/commands/main.go:5]: Hmm, I did it that way, but git still wants to see this as a new file... Maybe there are too many changes? Weird.
